compiler/circle-mpqsolver/src/core/SolverHooks.h

```diff
@@ -61,6 +61,15 @@ class SolverHooks
    */
   virtual void on_end_solver(const LayerParams &layers, const std::string &def_dtype,
                              float qerror) = 0;
```

Comment:
I think this should be replaced with new `on_end_solver`?

Reply:
> I think this should be replaced with new `on_end_solver`? Or at least TODO comment needs to be inserted.

ok.

Reply:
I've posted comment for now.

Reply:
@jinevening 
Please ignore comment above. I've merged these functions into single one
```
 virtual void on_end_solver(const LayerParams &layers, const std::string &def_dtype,
                             const std::string *def_granularity, const float *qerror);
```
so that unapplicable parameters are passed as nullptr.
Please let me know if this design is not satisfactory.

Reply:
@jinevening 
Can i try to split into mergeable PRs?

Reply:
@jinevening 
Please take another look.

Reply:
Sorry for late reply. I forgot this issue for a while (maybe missed when I came back from leave).

> Please let me know if this design is not satisfactory.
Or it's better to introduce some structure like Quantizer Context or don't use pointers (so that empty string and -1 error will mean unapplicability)?

Making a struct looks good to me. BTW, is it necessary to use hard-coded values like "channel" or "uint8"? Now those information is saved in Quantizer::Context.

Reply:
> Making a struct looks good to me. BTW, is it necessary to use hard-coded values like "channel" or "uint8"? Now those information is saved in Quantizer::Context.

Ahhh. I'll try to remove hard-coded values. Thank you.

Reply:
@jinevening 
done